packages/gatsby/src/utils/find-page-by-path.ts

```diff
@@ -0,0 +1,88 @@
+import { IGatsbyPage, IGatsbyState } from "../redux/types"
+import { match } from "@reach/router/lib/utils"
+
+export function findPageByPath(
+  state: IGatsbyState,
+  path: string,
+  fallbackTo404: boolean = false
+): IGatsbyPage | undefined {
+  const { pages } = state
+
+  // first check by exact path
+  let page = pages.get(path)
+  if (page) {
+    return page
+  }
+
+  if (path === ``) {
+    // from my tests I never was able to make request with
+    // completely empty pathname, but just for the sake
+    // of completeness - try available alternative
+    page = pages.get(`/`)
+    if (page) {
+      return page
+    }
+  }
+  // Gatsby doesn't allow for page path to be empty string,
+  // so skipping trying to get page for "" path if we can't
+  // find page for `/`
+  else if (page !== `/`) {
+    // check various trailing/leading slashes combinations
+    const hasLeadingSlash = path.startsWith(`/`)
+    const hasTrailingSlash = path.endsWith(`/`)
+    for (const leadingSlash of [true, false]) {
+      for (const trailingSlash of [true, false]) {
+        if (
+          leadingSlash === hasLeadingSlash &&
+          trailingSlash === hasTrailingSlash
+        ) {
+          // we checked this already
+          continue
+        }
+
+        let newPath = path
+
+        if (leadingSlash !== hasLeadingSlash) {
+          if (leadingSlash) {
+            newPath = `/` + newPath
+          } else {
+            newPath = newPath.substring(1)
+          }
+        }
+
+        if (trailingSlash !== hasTrailingSlash) {
+          if (trailingSlash) {
+            newPath = newPath + `/`
+          } else {
+            newPath = newPath.substring(0, newPath.length - 1)
+          }
+        }
+
+        page = pages.get(newPath)
+        if (page) {
+          return page
+        }
+      }
+    }
+  }
+
+  // we didn't find exact static page, time to check matchPaths
+  for (const [, page] of pages) {
+    if (page.matchPath && match(page.matchPath, path)) {
+      return page
+    }
+  }
+
+  if (fallbackTo404) {
+    page = findPageByPath(state, `/dev-404-page/`, false)
```

Comment:
Why prefer this? Should this not be condiitonal?

Reply:
By "this", do you mean `/dev-404-page/` over `/404.html` or something else?

If this is about above:
There is no correct order - `/dev-404-page/` _should_ exist only in `gatsby develop` case, but also it's really special cased in our develop runtime that when you navigate to _404_ page then we serve both dev-404 and "real 404" (if it exists) - that's what makes the button that shows real 404 instead of dev-404 work right now (not fan of the implementation, but also didn't want to expand the scope of streamlining that part).

I could make it conditional to only check dev-404-page in develop, but also this utility function would only be used in `gatsby develop` scenario and not the builds. Primary purpose of it is to be used in various request handlers - be it express handler (i.e. SSR in develop or triggering query on demand on fetches for `page-data` files - both not in this PR) or the websocket handlers (in this PR). I'm not sure if there is value in using this utility in builds ever - there we should always use canonical page paths to avoid uncertainty if path we use matches exactly what page we have. I can note this down in jsdoc description